codec-http2/src/main/java/io/netty/handler/codec/http2/ChildChannel.java

```diff
@@ -0,0 +1,252 @@
+    protected void doDisconnect() throws Exception {
+        doClose();
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+        closed = true;
+    }
+
+    @Override
+    protected void doDeregister() throws Exception {
+        // Just remove the shutdownHook as this Channel may be closed later or registered to another
+        // EventLoop
+        ((SingleThreadEventLoop) eventLoop().unwrap()).removeShutdownHook(shutdownHook);
+    }
+
+    @Override
+    protected void doBeginRead() throws Exception {
+        if (readInProgress) {
+            return;
+        }
+
+        if (inboundBuffer.isEmpty()) {
+            readInProgress = true;
+            return;
+        }
+
+        final InternalThreadLocalMap threadLocals = InternalThreadLocalMap.get();
+        final Integer stackDepth = threadLocals.localChannelReaderStackDepth();
+        if (stackDepth < MAX_READER_STACK_DEPTH) {
+            threadLocals.setLocalChannelReaderStackDepth(stackDepth + 1);
+            try {
+                readTask.run();
+            } finally {
+                threadLocals.setLocalChannelReaderStackDepth(stackDepth);
+            }
+        } else {
+            eventLoop().execute(readTask);
+        }
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw new ClosedChannelException();
+        }
+
+        EventLoop parentLoop = parent().eventLoop();
+
+        if (parentLoop.inEventLoop()) {
+            try {
+                for (;;) {
+                    Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    parentWrite(msg);
+                    in.remove();
+                }
+                parentWriteComplete();
+            } catch (Throwable t) {
+                parentExceptionCaught(t);
+            }
+        } else {
+            // Use a copy because the original msgs will be recycled by AbstractChannel.
+            final Object[] msgsCopy = new Object[in.size()];
+            for (int i = 0; i < msgsCopy.length; i ++) {
+                msgsCopy[i] = ReferenceCountUtil.retain(in.current());
+                in.remove();
+            }
+
+            parentLoop.execute(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        for (Object msg : msgsCopy) {
+                            parentWrite(msg);
+                            ReferenceCountUtil.release(msg);
+                        }
+                        parentWriteComplete();
+                    } catch (Throwable t) {
+                        parentExceptionCaught(t);
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Always called from the parent's event loop.
+     */
+    protected abstract void parentWrite(Object msg) throws Exception;
```

Comment:
Agreed that having a clear definition for what methods run in which eventloop/thread context (and I think you have done a good job with this).  However I'm not sure this has to dictate the method names.  I think most of Netty just documents these restrictions in the interface.  I think the whole child/parent channel construct is built on the assumption you are using the "parent" channel to actually do work, so I'm not sure if including it in the method name clarifies anything.  Maybe I just need to take a fresh look tomorrow.